src/lexer/regexFactory.ts

```diff
@@ -0,0 +1,98 @@
+import { escapeRegExp, sortByLengthDesc } from '../utils';
+
+export const NULL_REGEX = /(?!)/; // zero-width negative lookahead, matches nothing
+
+export const lineComment = (lineCommentTypes: string[]) =>
+  new RegExp(`(?:${lineCommentTypes.map(escapeRegExp).join('|')}.*?)(?:\r\n|\r|\n|$)`, 'u');
+
+export const operator = (monadOperators: string, polyadOperators: string[]) =>
+  new RegExp(
+    `${sortByLengthDesc(polyadOperators).map(escapeRegExp).join('|')}|` +
+      `[${monadOperators.split('').map(escapeRegExp).join('')}]`,
+    'u'
+  );
+
+// This enables the following string patterns:
+// 1. backtick quoted string using `` to escape
+// 2. square bracket quoted string (SQL Server) using ]] to escape
+// 3. double quoted string using "" or \" to escape, with optional prefix for format-specific strings
+// 4. single quoted string using '' or \' to escape, with optional prefix for format-specific strings
+// 8. PostgreSQL dollar-quoted strings (does not check for matching tags due to moo not allowing capturing groups)
+
+const stringPrefixList = ['[Nn]', '_utf8', 'U&', 'x', '[Rr]', '[Bb]', 'E'];
+const createStringPattern = (stringPrefixes: string) => ({
+  '``': '(?:`[^`]*(?:$|`))+',
+  '{}': '(?:\\{[^\\}]*(?:$|\\}))+',
+  '[]': '(?:\\[[^\\]]*(?:$|\\]))(\\][^\\]]*(?:$|\\]))*',
+  '""': `(?:${stringPrefixes}"[^"\\\\]*(?:\\\\.[^"\\\\]*)*(?:"|$))+`,
+  "''": `(?:${stringPrefixes}'[^'\\\\]*(?:\\\\.[^'\\\\]*)*(?:'|$))+`,
+  // '$$': '(?<tag>\\$\\w*\\$)[\\s\\S]*?(?:\\k<tag>|$)', // does not work with moo
```

Comment:
Why doesn't this work with Moo and what could be done about this?

Reply:
https://github.com/no-context/moo/pull/132
moo doesn't allow capturing groups since it tries to avoid parsing regex in general

Reply:
Seems to be a pretty serious limitation. This issue is there since 2019 and no progress has really happened about this in Moo since then :(

I see three possibilities:

- Drop support for `$$` strings from sql-formatter. Not really feasible.
- Attempt fixing this issue in Moo. Will take quite some time at very minimum. Might never happen.
- Use patched version of Moo (applying this back-reference support PR). Not future-proof. Better to be avoided.
- Don't use Moo. Find some other lexer or enhance our home-grown one.

Reply:
$$ strings do work, the issue is only for tagged $$ strings such as $xx$string_content$xx$
moo _can_ match a token of the form $xx$string_content$yy$, it just can't ensure that xx == yy without the backreference

we could either
- parse $xx$, string_content, $yy$ as 3 separate tokens + validate and join in post processor
- or just parse $xx$string_content$yy$ as is, and leave it up to the user to ensure xx == yy

Reply:
Well, the problem is, that these two approaches don't work.

- In first case the content between `$xx$..$xx$` is not necessarily valid SQL. For example one could have SQL like this: `SELECT $xx$some "content$xx$;`. This would get tokenized to `SELECT`, `$xx`, `some`, `"content$xx$;`.
- The second case will fail with a string like `$xx$ some prices: $18$42 $xx$`.